drv/i2c-devices/src/sbrmi.rs

```diff
@@ -0,0 +1,339 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Driver for AMD SB-RMI interface for AMD Milan.  This interface is both
+//! AMD- and Milan-specific (and in particular, note that the number of
+//! threads cannot exceed a 7-bit quantity in this processor generation).
+
+use crate::Validate;
+use drv_i2c_api::*;
+use ringbuf::*;
+use zerocopy::FromBytes;
+
+#[allow(dead_code)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Register {
+    Revision,
+    Control,
+    Status,
+    ReadSize,
+    ThreadNumber,
+    Enabled { base: u8, offset: u8 },
+    Alert { base: u8, offset: u8 },
+}
+
+impl From<Register> for u8 {
+    fn from(reg: Register) -> Self {
+        match reg {
+            Register::Revision => 0x0,
+            Register::Control => 0x1,
+            Register::Status => 0x2,
+            Register::ReadSize => 0x3,
+            Register::ThreadNumber => 0x41,
+            Register::Enabled { base, offset } => base + offset,
+            Register::Alert { base, offset } => base + offset,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum StatusCode {
+    Success,
+    CommandTimeout,
+    WarmReset,
+    UnknownCommandFormat,
+    InvalidReadLength,
+    ExcessiveData,
+    InvalidThread,
+    UnsupportedCommand,
+    CommandAborted,
+    Unknown(u8),
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Default)]
+pub struct CpuidResult {
+    pub eax: u32,
+    pub ebx: u32,
+    pub ecx: u32,
+    pub edx: u32,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Error {
+    BadRegisterRead { reg: Register, code: ResponseCode },
+    BadThreadId,
+    BadCpuidInput,
+    BadCpuidLength { length: u8 },
+    BadCpuidRead { code: ResponseCode },
+    CpuidFailed { code: StatusCode },
+    BadRdmsrLength { length: u8 },
+    BadRdmsr { code: ResponseCode },
+    RdmsrFailed { code: StatusCode },
+}
+
+impl From<Error> for ResponseCode {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::BadRegisterRead { code, .. } => code,
+            Error::BadCpuidRead { code } => code,
+            _ => ResponseCode::BadResponse,
```

Comment:
Doing a quick audit, it seems like different drivers do many different things when this conversion fails:
- `bmr491` (and many others) will panic
- `nvme_bmc` (and a few others) will return `ResponseCode::BadDeviceState`
- Here, we're returning `ResponseCode::BadResponse`

Philosophically, I think this should be `TryFrom`, but would require modifying I2C validation codegen.  I opened #1161 to track this.

Reply:
Thanks for filing that!